### PR TITLE
[Enhance] Speed up evaluation on waymo

### PR DESCRIPTION
**Pull request by JingweiZhang12**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Speed up evaluation on waymo through three aspects:
1. not converting to kitti-format
2. reading metainfo from the preprocessed file rather than from original files online.
3. using multi-thread waymo evaluation toolkit: https://github.com/Abyssaledge/waymo-open-dataset-master

The ablation of evaluation time on Waymo for Lidar-based methods:
|  Not to KITTI format  |  Offline metainfo  |   Multi-thread eval toolkit |  time(min)  |
| :--------------------: | :---------------------: | :------------------------------: | :------: |
|     |     |   |  30  |
| √  |    |    |  25.5  |
| √  | √   |   |  18.2  |
| √  | √   | √  |  5  |

## BC-breaking (Optional)

Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
3. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
5. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
6. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: docs/en/user_guides/dataset_prepare.md [hunk 1]
```diff
@@ -110,7 +110,25 @@ Download Waymo open dataset V1.2 [HERE](https://waymo.com/open/download/) and it
...
+      ann_file='./data/waymo/kitti_format/waymo_infos_val.pkl',
+      waymo_bin_file='./data/waymo/waymo_format/gt.bin',
+      data_root='./data/waymo/waymo_format',
+      file_client_args=file_client_args,
+      convert_kitti_format=True,
+      idx2metainfo='data/waymo/waymo_format/idx2metainfo.pkl'
+      )
+  ```
+
+  Experimently, this will reduce evalaution time by around 6 min. Now, this trick is only used in methods based in point cloud.
```
Comment (ZCMax) [inline docs/en/user_guides/dataset_prepare.md:131]: The whole evaluation process on Waymo costs around 6 min with 8 A100 GPUs? We need to specify the equipment.

Comment (ZCMax) [reply docs/en/user_guides/dataset_prepare.md:131]: Now this trick is only used in point cloud detection model

File: mmdet3d/evaluation/functional/waymo_utils/prediction_to_waymo.py [hunk 2]
```diff
@@ -5,63 +5,80 @@
...
+        results (list[dict]): Predictions in KITTI format.
         waymo_tfrecords_dir (str): Directory to load waymo raw data.
         waymo_results_save_dir (str): Directory to save converted predictions
             in waymo format (.bin files).
         waymo_results_final_path (str): Path to save combined
             predictions in waymo format (.bin file), like 'a/b/c.bin'.
         prefix (str): Prefix of filename. In general, 0 for training, 1 for
             validation and 2 for testing.
+        classes (dict): A list of class name.
         workers (str): Number of parallel processes.
```
Comment (ZCMax) [inline mmdet3d/evaluation/functional/waymo_utils/prediction_to_waymo.py:43]: Defaults to 2.

File: mmdet3d/evaluation/functional/waymo_utils/prediction_to_waymo.py [hunk 3]
```diff
@@ -5,63 +5,80 @@
...
+        file_client_args (str): File client for reading gt in waymo format.
```
Comment (ZCMax) [inline mmdet3d/evaluation/functional/waymo_utils/prediction_to_waymo.py:44]: Defaults to xxxx

File: mmdet3d/evaluation/functional/waymo_utils/prediction_to_waymo.py [hunk 4]
```diff
@@ -5,63 +5,80 @@
...
+class Prediction2Waymo(object):
+    """Predictions to Waymo converter. The format of prediction results could
+    be original format or kitti-format.
 
     This class serves as the converter to change predictions from KITTI to
     Waymo format.
 
     Args:
-        kitti_result_files (list[dict]): Predictions in KITTI format.
+        results (list[dict]): Predictions in KITTI format.
```
Comment (ZCMax) [inline mmdet3d/evaluation/functional/waymo_utils/prediction_to_waymo.py:34]: revome `in KITTI format`

**Comment by xyupeng** [comment]
Hi @JingweiZhang12 ,

Thanks for your fast evaluation code. Can you kindly share the file `idx2metainfo.pkl` required for fast evaluation? I didn't find it in this commit.

File: docs/en/user_guides/dataset_prepare.md (hunk 1 above)
Comment (JingweiZhang12) [reply docs/en/user_guides/dataset_prepare.md:131]: > The whole evaluation process on Waymo costs around 6 min with 8 A100 GPUs? We need to specify the equipment.

It reduces time **by** 6 min. But the file reading is not related to GPU and it's related to file IO speed, CPUs and other staffs. So I ignore it.

**Comment by JingweiZhang12** [comment]
> Hi @JingweiZhang12 ,
> 
> Thanks for your fast evaluation code. Can you kindly share the file `idx2metainfo.pkl` required for fast evaluation? I didn't find it in this commit.

Hi, @xyupeng, the link will come soon.

File: mmdet3d/evaluation/functional/waymo_utils/prediction_to_waymo.py [hunk 5]
```diff
@@ -232,11 +274,99 @@ def convert_one(self, file_idx):
                     'wb') as f:
                 f.write(objects.SerializeToString())
 
+    def convert_one_fast(self, res_index: int):
+        """_summary_
```
Comment (ZCMax) [inline mmdet3d/evaluation/functional/waymo_utils/prediction_to_waymo.py:278]: What does `_summary_` mean?

File: mmdet3d/evaluation/functional/waymo_utils/prediction_to_waymo.py [hunk 6]
```diff
@@ -232,11 +274,99 @@ def convert_one(self, file_idx):
...
+                join(self.waymo_results_save_dir, f'{sample_idx}.bin'),
+                'wb') as f:
+            f.write(objects.SerializeToString())
+
+    def parse_objects_from_origin(self, result: dict, contextname: str,
+                                  timestamp: str) -> Objects:
+        """_summary_
+
+        Args:
+            result (dict): _description_
```
Comment (ZCMax) [inline mmdet3d/evaluation/functional/waymo_utils/prediction_to_waymo.py:303]: check docstrings

File: mmdet3d/evaluation/metrics/waymo_metric.py [hunk 7]
```diff
@@ -54,6 +54,10 @@ class WaymoMetric(KittiMetric):
             from different ranks during distributed training. Must be 'cpu' or
             'gpu'. Defaults to 'cpu'.
         file_client_args (dict): file client for reading gt in waymo format.
```
Comment (ZCMax) [inline mmdet3d/evaluation/metrics/waymo_metric.py:56]: Defaults to xxxx

File: mmdet3d/evaluation/metrics/waymo_metric.py [hunk 8]
```diff
@@ -54,6 +54,10 @@ class WaymoMetric(KittiMetric):
...
+        idx2metainfo (Optional[str], optional): The file path of the metainfo
+            in waymmo. It stores the mapping from sample_idx to metainfo.
+            The metainfo must contain the keys: 'idx2contextname' and
+            'idx2timestamp'.
```
Comment (ZCMax) [inline mmdet3d/evaluation/metrics/waymo_metric.py:60]: Defaults to None

**Comment by xyupeng** [comment]
I have a question: is `Prediction2Waymo` based on the new coordinate system in `mmdet3d>=1.0` ? e.g., the new coordinate in `LiDARInstance3DBoxes`. Because I'm working on `mmdet3d==0.17`, but I still want to use fast waymo evalution.

**Comment by xyupeng** [comment]
> 

Thank so much!

**Comment by JingweiZhang12** [comment]
> I have a question: is `Prediction2Waymo` based on the new coordinate system in `mmdet3d>=1.0` ? e.g., the new coordinate in `LiDARInstance3DBoxes`. Because I'm working on `mmdet3d==0.17`, but I still want to use fast waymo evalution.

**Comment by JingweiZhang12** [comment]
> I have a question: is `Prediction2Waymo` based on the new coordinate system in `mmdet3d>=1.0` ? e.g., the new coordinate in `LiDARInstance3DBoxes`. Because I'm working on `mmdet3d==0.17`, but I still want to use fast waymo evalution.

Yes,  it's  baed on the new coordinate system. If you use the old coordinate system, you should note that:
```
Our LiDAR coordinate system has two changes:
The yaw angle is defined to be right-handed instead of left-handed for consistency;
The box dimensions are  instead of , since  corresponds to  and  corresponds to  in KITTI.
```
Details can be seen https://mmdetection3d.readthedocs.io/en/latest/tutorials/coord_sys_tutorial.html

**Comment by xyupeng** [comment]
Sorry for choosing the wrong place to comment. But I went through the data pipeline for waymo in mmdet3d 0.17.
I find that the dimensions in `gt_bboxes_3d[:, 3:6]` are `wlh` (width, length, height), but not `lwh` as defined in waymo label proto (i.e., waymo label proto defines `xyz` correspond to `lwh`).

The pipeline is:
1. kitti_format/training/label_all/0000000.txt: `hwl`
2. kitti_format/training/waymo_infos_train.pkl: info['annos']['dimensions']: `lhw`
3. WaymoDataset.get_ann_info(): ann_results['gt_bboxes_3d'][:, 3:6]: `wlh` (from `CameraInstance3DBoxes` to `LiDARInstance3DBoxes` )

So the final input `gt_bboxes_3d` has dimensions `wlh`? Is this the expected behavior instead of `lwh`?

**Comment by JingweiZhang12** [comment]
> Sorry for choosing the wrong place to comment. But I went through the data pipeline for waymo in mmdet3d 0.17. I find that the dimensions in `gt_bboxes_3d[:, 3:6]` are `wlh` (width, length, height), but not `lwh` as defined in waymo label proto (i.e., waymo label proto defines `xyz` correspond to `lwh`).
> 
> The pipeline is:
> 
> 1. kitti_format/training/label_all/0000000.txt: `hwl`
> 2. kitti_format/training/waymo_infos_train.pkl: info['annos']['dimensions']: `lhw`
> 3. WaymoDataset.get_ann_info(): ann_results['gt_bboxes_3d'][:, 3:6]: `wlh` (from `CameraInstance3DBoxes` to `LiDARInstance3DBoxes` )
> 
> So the final input `gt_bboxes_3d` has dimensions `wlh`? Is this the expected behavior instead of `lwh`?

The old lidar cooridnate system is 'wlh'. https://mmdetection3d.readthedocs.io/en/latest/tutorials/coord_sys_tutorial.html

**Comment by xyupeng** [comment]
> > Sorry for choosing the wrong place to comment. But I went through the data pipeline for waymo in mmdet3d 0.17. I find that the dimensions in `gt_bboxes_3d[:, 3:6]` are `wlh` (width, length, height), but not `lwh` as defined in waymo label proto (i.e., waymo label proto defines `xyz` correspond to `lwh`).
> > The pipeline is:
> > 
> > 1. kitti_format/training/label_all/0000000.txt: `hwl`
> > 2. kitti_format/training/waymo_infos_train.pkl: info['annos']['dimensions']: `lhw`
> > 3. WaymoDataset.get_ann_info(): ann_results['gt_bboxes_3d'][:, 3:6]: `wlh` (from `CameraInstance3DBoxes` to `LiDARInstance3DBoxes` )
> > 
> > So the final input `gt_bboxes_3d` has dimensions `wlh`? Is this the expected behavior instead of `lwh`?
> 
> The old lidar coordinate system (before version 1.0) has dimention `wlh`. https://mmdetection3d.readthedocs.io/en/latest/tutorials/coord_sys_tutorial.html

Got it. Thanks so much.

**Comment by ZCMax** [comment]
Please merge the latest dev-1.x branch and resolve conflicts

File: docs/en/user_guides/dataset_prepare.md [hunk 9]
```diff
@@ -110,7 +110,25 @@ Download Waymo open dataset V1.2 [HERE](https://waymo.com/open/download/) and it
...
+      ann_file='./data/waymo/kitti_format/waymo_infos_val.pkl',
+      waymo_bin_file='./data/waymo/waymo_format/gt.bin',
+      data_root='./data/waymo/waymo_format',
+      file_client_args=file_client_args,
+      convert_kitti_format=True,
+      idx2metainfo='data/waymo/waymo_format/idx2metainfo.pkl'
+      )
+  ```
+
+  Experimently, this will reduce evalaution time by around 6 min. Now, this trick is only used in point cloud detection model.
```
Comment (Tai-Wang) [inline docs/en/user_guides/dataset_prepare.md:131]: Quantitatively/Empirically/Experimentally

Comment (Tai-Wang) [reply docs/en/user_guides/dataset_prepare.md:131]: this trick is only used for LiDAR-based detection methods.

**Review by Tai-Wang — COMMENTED** [review]
Convert to KITTI format for camera-only models while not for LiDAR-based models by default. Please provide a basic benchmark table for reference.

**Pull request merged**